completion_aggregator/aggregator.py

```diff
@@ -0,0 +1,173 @@
+        tick = service_start_time
+        i = 0
+        while tick < service_end_time:
+            i += 1
+            if i % 1000 == 0:
+                self.perform_cleanup()
+            self.perform_service(tick)
+            nexttick = tick + self.period
+            iteration_end_time = timezone.now()
+            if iteration_end_time > nexttick:
+                if self.period.total_seconds() > 0:
+                    log.warning(
+                        "%s iteration took longer than configured period. Took %s; Period: %s",
+                        self.service_name,
+                        iteration_end_time - tick,
+                        self.period,
+                    )
+                tick = iteration_end_time
+            else:
+                tick = nexttick
+                sleep_time = (tick - iteration_end_time).total_seconds()
+                time.sleep(sleep_time)
+
+        self.perform_cleanup()
+
+    def run_once(self):
+        """
+        Run the task once, then clean up and exit.
+        """
+        self.perform_service(timezone.now())
+        self.perform_cleanup()
+
+    def perform_service(self, tick):
+        """
+        Do the task we're here to do.
+        """
+        raise NotImplementedError
+
+    def perform_cleanup(self):
+        """
+        Perform a cleanup task if needed.
+        """
+        pass
+
+
+class AggregatorService(BasePeriodicService):
+    """
+    Enqueues tasks to reaggregate modified completions.
+
+    When blocks are completed, they mark themselves as stale.  This service
+    collects all stale blocks for each enrollment, and enqueues a single
+    recalculation of all aggregators containing those stale blocks.
+    """
+
+    service_name = "Aggregator Service"
+    enrollment_tuple = collections.namedtuple('enrollment_tuple', ['username', 'course_key'])
+
+    def perform_service(self, tick):
+        """
+        Collect stale blocks, and enqueue new aggregation tasks.
+        """
+        with transaction.atomic():
+            stale_queryset = models.StaleCompletion.objects.select_for_update().filter(resolved=False)
+            stale_blocks = collections.defaultdict(set)
+            forced_updates = set()
+            for stale in stale_queryset:
+                enrollment = self.enrollment_tuple(
+                    username=stale.username,
+                    course_key=stale.course_key,
+                )
+                if stale.block_key is None:
+                    stale_blocks[enrollment] = utils.BagOfHolding()
+                else:
+                    stale_blocks[enrollment].add(stale.block_key)
+                if stale.force:
+                    forced_updates.add(enrollment)
+
+            log.info("Performing aggregation update for %s user enrollments", len(stale_blocks))
+            for enrollment in stale_blocks:
+                if isinstance(stale_blocks[enrollment], utils.BagOfHolding):
```

Comment:
I really want to combine this condition with the next one by making `len(BagOfHolding()) == float('inf')`..  But `__len__()` must return an integer, and infinity won't fit in a lowly integer 😦

Reply:
I suppose I could set len(BagOfHolding) to something absurdly large.  I'd use a MAXINT value, but python doesn't really have that concept with its boundless integers.  2**1024 should be big enough that it will never conflict with an actual set, but I'm not sure if having an arbitrary big len is better or worse than adding an extra condition.

Reply:
I think the extra condition you used here is the clearest way to make sure.. I just had a wish 😄